src/nupic/algorithms/AnomalyLikelihood.cpp

```diff
@@ -0,0 +1,592 @@
+    name(name),mean(mean), variance(variance), stdev(stdev) {}
+  string name;
+  Real32 mean;
+  Real32 variance;
+  Real32 stdev;
+};
+
+class AnomalyLikelihood {
+public:
+    //Constructors
+
+  /**
+      NOTE: Anomaly likelihood scores are reported at a flat 0.5 for
+    learningPeriod + estimationSamples iterations.
+
+    claLearningPeriod and learningPeriod are specifying the same variable,
+    although claLearningPeriod is a deprecated name for it.
+
+    @param learningPeriod (claLearningPeriod: deprecated) - (int) the number of
+      iterations required for the algorithm to learn the basic patterns in the
+      dataset and for the anomaly score to 'settle down'. The default is based
+      on empirical observations but in reality this could be larger for more
+      complex domains. The downside if this is too large is that real anomalies
+      might get ignored and not flagged.
+
+    @param estimationSamples - (int) the number of reasonable anomaly scores
+      required for the initial estimate of the Gaussian. The default of 100
+      records is reasonable - we just need sufficient samples to get a decent
+      estimate for the Gaussian. It's unlikely you will need to tune this since
+      the Gaussian is re-estimated every 10 iterations by default.
+
+    @param historicWindowSize - (int) size of sliding window of historical
+      data points to maintain for periodic reestimation of the Gaussian. Note:
+      the default of 8640 is based on a month's worth of history at 5-minute
+      intervals.
+
+    @param reestimationPeriod - (int) how often we re-estimate the Gaussian
+      distribution. The ideal is to re-estimate every iteration but this is a
+      performance hit. In general the system is not very sensitive to this
+      number as long as it is small relative to the total number of records
+      processed.
+
+  **/
+    AnomalyLikelihood(UInt learningPeriod=288, UInt estimationSamples=100, UInt historicWindowSize=8640, UInt reestimationPeriod=100, UInt aggregationWindow=10) :
+    learningPeriod(learningPeriod),
+    reestimationPeriod(reestimationPeriod),
+    averagedAnomaly(aggregationWindow) {
+        iteration = 0;
+        probationaryPeriod = learningPeriod+estimationSamples;
+        assert(historicWindowSize >= estimationSamples); // cerr << "estimationSamples exceeds historicWindowSize";
+        assert(aggregationWindow < reestimationPeriod && reestimationPeriod < historicWindowSize);
+        
+        runningAverageAnomalies.set_capacity(historicWindowSize); 
+        runningLikelihoods.set_capacity(historicWindowSize);
+        runningRawAnomalyScores.set_capacity(historicWindowSize);
+        runningRawValues.set_capacity(historicWindowSize);
+        assert(runningLikelihoods.capacity() == historicWindowSize);
+    }
+
+    
+    Real anomalyProbability(Real rawValue, Real anomalyScore, int timestamp=-1) { //TODO "rawValue" could be typed to T rawValue; //TODO is rawValue ever used?
```

Comment:
TODO: `rawValue` could be templated (T), not a priority now. 
FIXME: is the rawValue ever used? Does not seem so in my implementation.